proto/src/test/java/com/google/gson/protobuf/functional/ProtosWithAnnotationsAndJsonNamesTest.java

```diff
@@ -0,0 +1,206 @@
+          .create();
+  private static final Gson GSON_WITH_JSON_NAME =
+      new GsonBuilder()
+          .registerTypeHierarchyAdapter(
+              GeneratedMessage.class,
+              ProtoTypeAdapter.newBuilder().setShouldUseJsonNameFieldOption(true).build())
+          .create();
+  private static final Gson GSON_WITH_SERIALIZED_NAME_AND_JSON_NAME =
+      new GsonBuilder()
+          .registerTypeHierarchyAdapter(
+              GeneratedMessage.class,
+              ProtoTypeAdapter.newBuilder()
+                  .addSerializedNameExtension(Annotations.serializedName)
+                  .setShouldUseJsonNameFieldOption(true)
+                  .build())
+          .create();
+
+  private static final Map<Gson, String> JSON_OUTPUTS =
+      Map.of(
+          GSON_PLAIN,
+              "{\"neither\":\"xxx\",\"jsonNameOnly\":\"yyy\",\"annotationOnly\":\"zzz\",\"both\":\"www\"}",
+          GSON_WITH_JSON_NAME,
+              "{\"neither\":\"xxx\",\"aaa\":\"yyy\",\"annotationOnly\":\"zzz\",\"ccc\":\"www\"}",
+          GSON_WITH_SERIALIZED_NAME,
+              "{\"neither\":\"xxx\",\"jsonNameOnly\":\"yyy\",\"bbb\":\"zzz\",\"ddd\":\"www\"}",
+          GSON_WITH_SERIALIZED_NAME_AND_JSON_NAME,
+              "{\"neither\":\"xxx\",\"aaa\":\"yyy\",\"bbb\":\"zzz\",\"ddd\":\"www\"}");
+
+  private static final ProtoWithAnnotationsAndJsonNames PROTO =
+      ProtoWithAnnotationsAndJsonNames.newBuilder()
+          .setNeither("xxx")
+          .setJsonNameOnly("yyy")
+          .setAnnotationOnly("zzz")
+          .setBoth("www")
+          .build();
+
+  @Test
+  public void testProtoWithAnnotationsAndJsonNames_basicConversions() {
+    JSON_OUTPUTS.forEach(
+        (gson, json) -> {
+          assertThat(gson.fromJson(json, ProtoWithAnnotationsAndJsonNames.class)).isEqualTo(PROTO);
+          assertThat(gson.toJson(PROTO)).isEqualTo(json);
+        });
+  }
+
+  @Test
+  public void testProtoWithAnnotationsAndJsonNames_basicRoundTrips() {
+    JSON_OUTPUTS.forEach(
+        (gson, json) -> {
+          assertThat(roundTrip(gson, gson, json)).isEqualTo(json);
+          assertThat(roundTrip(gson, gson, PROTO)).isEqualTo(PROTO);
+        });
+  }
+
+  @Test
+  public void testProtoWithAnnotationsAndJsonNames_unannotatedField() {
+    ProtoWithAnnotationsAndJsonNames proto =
+        ProtoWithAnnotationsAndJsonNames.newBuilder().setNeither("zzz").build();
+    String json = "{\"neither\":\"zzz\"}";
+
+    for (Gson gson1 : JSON_OUTPUTS.keySet()) {
+      for (Gson gson2 : JSON_OUTPUTS.keySet()) {
+        // all configs should match with each other in how they serialize this proto, and they
+        // should be able to deserialize any other config's serialization of the proto back to its
+        // original form
+        assertThat(gson1.toJson(proto)).isEqualTo(gson2.toJson(proto));
+        assertThat(roundTrip(gson1, gson2, proto)).isEqualTo(proto);
+        assertThat(roundTrip(gson2, gson1, proto)).isEqualTo(proto);
```

Comment:
Not that it matters, but strictly speaking this assertion is redundant since if `gson1 = a` and `gson2 = b`, there will be another combination of the nested loops where `gson1 = b` and `gson2 = a`, and of course likewise for line 123.